rfcs/20200810-tf-internal-api.md

```diff
@@ -0,0 +1,118 @@
+
+Keras team is trying to split its code into a [standalone Github repository](https://github.com/tensorflow/community/blob/master/rfcs/20200205-standalone-keras-repository.md). 
+One of the key actions before the repository split is that Keras code has to 
+rely on the public TF API only, to avoid any future breakages due to private 
+method/class change.
+
+Historically, Keras was an internal component of TF, and relies heavily on a lot
+of private TF utilities and functionalities that are not part of the public API. 
+Exposing those functions into public APIs will be very confusing to the end 
+user, since those functions are useful to framework-builders, but not typical 
+TensorFlow users. In addition, these utilities would be overly restricted by the 
+standard TensorFlow API contract, as we would like to change/remove them more 
+often than the existing core API.
+
+
+## Design Proposal
+
+Add a new namespace "tensorflow.internal" to host the framework building/testing 
+related APIs. <b>(See alternative naming in the sections below, the naming of 
+the namespace will be discussed in the design meeting.)</b>
+
+### Backward compatibility
+The contract for all the APIs under this namespace is looser compared to core TF
+API. It will remain backward compatible for at least 1 minor TF release. 
+
+This means for any API that is added in 2.x release, it will remain the same in 
+2.x+1. If we choose to remove the same API, we will mark it as deprecated in 
+2.x+1, and delete it at 2.x+2 release. TensorFlow is released every 3 months, 
+and this will give enough time for all the clients to change to the new 
+API/alternative.
+
+Any deprecation and backward incompatible API change will be explicitly called 
+out in TF release notes. This applies to both core API and "internal" API.
+
+### Acceptance Criteria
+The candidate of the "internal" API should:
+
+1. Does NOT fit for core TF API, otherwise it should be exposed as core TF API.
+1. Are effectively required to build/test/deploy a framework or library on top 
+   of TensorFlow, or to integrate such a framework into TensorFlow's API.
+1. Mature enough and won't change behavior/interface for every release. 
+1. Non-trivial, otherwise the function should be copied to the client side.
+1. Widely used in the implementation of TF public APIs (i.e. new functionality 
+   isn’t immediately added to the tf.internal namespace)
+1. Has at least two downstream libraries which are known to need it.
+
+TF API owners will review the new API proposal, and follow the existing review 
+process for core TF API.
+
+
+### Documentation and Docstring
+The "internal" API should have the same style and standard as the core 
+TensorFlow API, which is documented [here](https://github.com/tensorflow/community/blob/master/governance/api-reviews.md#docstrings). 
+We should explicitly list out the difference between "internal" API and core
+API, and also choose a different place on tensorflow.org so that existing user 
```

Comment:
@markdaoust and @yashk2810, do u have any suggestion here for a place on tensorflow.org to host this internal API?

Reply:
If its in the TF package then it will be on the API docs page here: https://www.tensorflow.org/api_docs/python/tf?version=nightly with a `tf.internal` namespace as all the others in the left TOC.

Does that work?

Reply:
How many symbols are going to be left in this? I'd still prefer we don't have this publicly.

Reply:
Yes, we can block this from being displayed on tensorflow.org if that's the way we want to go.

Reply:
I think we can block it in the beginning, since it will cause confusion for end user. We might need to choose a different place for it, targeting framework builder only.